large `List` causes assertion failed in interflow

This was encountered in https://github.com/typelevel/idna4s/pull/64, which uses source generation to implement the UTS46 standard for IDNA, which often build large collections.

Specifically, https://github.com/typelevel/idna4s/pull/64/commits/f996802220673f9b430f9fa98042f13875ad79bb causes the `nativeLink` step to crash due to this assertion.

https://github.com/scala-native/scala-native/blob/51b6b29cb8175194126208ae172d0b08f03cbc56/tools/src/main/scala/scala/scalanative/interflow/UseDef.scala#L89

### Minimizer

https://gist.github.com/armanbilge/2032a9623b7acfa36771e5165f30682d

### Output

<details>

```
scala-cli --native debug.scala 
Compiling project (Scala 3.2.1, Scala Native)
Compiled project (Scala 3.2.1, Scala Native)
Picked up JAVA_TOOL_OPTIONS:  -Xmx3489m
[info] Linking (10723 ms)
[info] Discovered 658 classes and 3684 methods
[info] Optimizing (debug mode) (10101 ms)
Exception in thread "main" java.lang.AssertionError: assertion failed
        at scala.Predef$.assert(Predef.scala:208)
        at scala.scalanative.interflow.UseDef$.enterInst$1(UseDef.scala:89)
        at scala.scalanative.interflow.UseDef$.$anonfun$apply$5(UseDef.scala:116)
        at scala.collection.mutable.UnrolledBuffer$Unrolled.foreach(UnrolledBuffer.scala:242)
        at scala.collection.mutable.UnrolledBuffer.foreach(UnrolledBuffer.scala:148)
        at scala.scalanative.interflow.UseDef$.$anonfun$apply$3(UseDef.scala:114)
        at scala.scalanative.interflow.UseDef$.$anonfun$apply$3$adapted(UseDef.scala:112)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
        at scala.scalanative.interflow.UseDef$.apply(UseDef.scala:112)
        at scala.scalanative.interflow.UseDef$.eliminateDeadCode(UseDef.scala:155)
        at scala.scalanative.codegen.Lower$Impl.onInsts(Lower.scala:180)
        at scala.scalanative.nir.Transform.onDefn(Transform.scala:18)
        at scala.scalanative.nir.Transform.onDefn$(Transform.scala:8)
        at scala.scalanative.codegen.Lower$Impl.super$onDefn(Lower.scala:84)
        at scala.scalanative.codegen.Lower$Impl.$anonfun$onDefn$1(Lower.scala:84)
        at scala.scalanative.util.ScopedVar$.scoped(ScopedVar.scala:43)
        at scala.scalanative.codegen.Lower$Impl.onDefn(Lower.scala:84)
        at scala.scalanative.codegen.Lower$Impl.$anonfun$onDefns$1(Lower.scala:72)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
        at scala.scalanative.codegen.Lower$Impl.onDefns(Lower.scala:63)
        at scala.scalanative.codegen.Lower$.apply(Lower.scala:23)
        at scala.scalanative.codegen.CodeGen$.$anonfun$lower$2(CodeGen.scala:37)
        at scala.collection.parallel.AugmentedIterableIterator.map2combiner(RemainsIterator.scala:116)
        at scala.collection.parallel.AugmentedIterableIterator.map2combiner$(RemainsIterator.scala:113)
        at scala.collection.parallel.immutable.ParHashMap$ParHashMapIterator.map2combiner(ParHashMap.scala:80)
        at scala.collection.parallel.ParIterableLike$Map.leaf(ParIterableLike.scala:1064)
        at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:53)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:67)
        at scala.collection.parallel.Task.tryLeaf(Tasks.scala:56)
        at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:50)
        at scala.collection.parallel.ParIterableLike$Map.tryLeaf(ParIterableLike.scala:1061)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal(Tasks.scala:170)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal$(Tasks.scala:157)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:440)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:150)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:149)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
        at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Error: scala.build.errors.ScalaNativeBuildError: Error compiling with Scala Native
For more details, please see '/workspace/sandbox/.scala-build/stacktraces/1673202207-1552726076556393695.log'
```

</details>

--- a/tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala
+++ b/tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala
@@ -12,6 +12,12 @@ final class MergeProcessor(
     doInline: Boolean,
     eval: Eval
 )(implicit linked: linker.Result) {
+  import MergeProcessor.MergeBlockOffset
+  assert(
+    insts.length < MergeBlockOffset,
+    s"Too big function, ${insts.length} instructions, max allowed ${MergeBlockOffset}"
+  )
+
   val offsets: Map[Local, Int] =
     insts.zipWithIndex.collect {
       case (Inst.Label(local, _), offset) =>
@@ -26,10 +32,13 @@ final class MergeProcessor(
   def findMergeBlock(name: Local): MergeBlock = {
     def newMergeBlock = {
       val label = insts(offsets(name)).asInstanceOf[Inst.Label]
-      new MergeBlock(label, Local(blockFresh().id * 10000))
+      this.newMergeBlock(label)
     }
     blocks.getOrElseUpdate(name, newMergeBlock)
   }
+
+  private def newMergeBlock(label: Inst.Label): MergeBlock =
+    new MergeBlock(label, Local(blockFresh().id * MergeBlockOffset))
 
   def merge(
       block: MergeBlock
@@ -439,8 +448,7 @@ final class MergeProcessor(
         Val.Local(syntheticFresh(), Sub.lub(tys, Some(retTy)))
       val syntheticLabel =
         Inst.Label(syntheticFresh(), Seq(syntheticParam))
-      val resultMergeBlock =
-        new MergeBlock(syntheticLabel, Local(blockFresh().id * 10000))
+      val resultMergeBlock = newMergeBlock(syntheticLabel)
       blocks(syntheticLabel.name) = resultMergeBlock
       orderedBlocks += resultMergeBlock
 
@@ -472,6 +480,8 @@ final class MergeProcessor(
 
 object MergeProcessor {
   case object Restart extends Exception with scala.util.control.NoStackTrace
+
+  private val MergeBlockOffset = 1000000L
 
   def fromEntry(
       insts: Array[Inst],

--- a/tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala
+++ b/tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala
@@ -481,6 +481,11 @@ final class MergeProcessor(
 object MergeProcessor {
   case object Restart extends Exception with scala.util.control.NoStackTrace
 
+  /* To mitigate risk of duplicated ids each merge block uses a dedicated
+   *  namespace. Translation to the new namespace is performed by multiplicating
+   *  id by value of MergeBlockOffset. This adds a restiction for maximal number
+   *  of instructions within a function to no larger then value of MergeBlockOffset.
+   */
   private val MergeBlockOffset = 1000000L
 
   def fromEntry(

